### DISCUSSION TITLE
Is there a way to drop build number from versions?

### DISCUSSION THREAD

**Discussion by jrouly** [projen/projen#2769] [Q&A] [answered]
I'm back with another version question!

Say my project is tagged `0.0.7`. If I run `PRERELEASE=foo projen bump`, that will yield a version bump `from 0.0.7 to 0.0.8-foo.0`.

Is there any way to drop the `.0` suffix?

The reason is, when publishing Maven artifacts, I've got `PRERELEASE=-SNAPSHOT`, but the artifact that gets built ends in `-SNAPSHOT.0` which is not being recognized by maven as a snapshot version.

**Comment by mrgrain** [comment, marked as answer]
projen currently uses [standard-version](https://github.com/conventional-changelog/standard-version) under hood. It doesn't seem to support that out-of-the-box.

**Reply by jrouly** [reply]
Okay - that was my understanding as well but wanted to make sure I wasn't missing something. Seems this precludes using jsii to publish maven snapshots of projen artifacts.

**Reply by mrgrain** [reply]
Yeah that might be the case. But I'd also guess that the tool was mostly selected from an node-ecosystem perspective.

**Reply by jrouly** [reply]
Totally makes sense. I'm only running into these issues now that I'm attempting to publish Python and JVM artifacts using jsii. Otherwise the workflow has been working great staying on the node ecosystem.

**Reply by mrgrain** [reply]
Do you have any references for the Maven snapshot issue? Not entirely sure yet what the solution would be (or where it should be implemented) but this does sound like something the toolchain (and I'm including projen AND jsii here) should support.

Feel free to open an issue as well!

**Reply by jrouly** [reply]
The closest I can find to official documentation from maven is [this example](https://maven.apache.org/guides/getting-started/#what-is-a-snapshot-version). Mostly it's based on my experience working with `mvn deploy`.

When I manually set my projen package version to `0.0.8-SNAPSHOT` vs `0.0.8-SNAPSHOT.0` and run `npx projen compile`, then `mvn deploy -f *.pom` in the resulting java package, I get different behaviors from maven.
- with the build number (`.0`) the artifact is interpreted as a release and my release repository is selected for publication
- without the build number, the artifact is interpreted as a snapshot and my snapshot repository is selected for publication

It looks like there are currently two related issues upstream in standard-version ([link](https://github.com/conventional-changelog/standard-version/issues/749) and [link](https://github.com/conventional-changelog/standard-version/issues/408)) and both seem related to maven-ecosystem build problems. Unfortunately it seems like that project has been deprecated and these issues aren't receiving attention :(

**Reply by mrgrain** [reply]
Yeah that makes a lot of sense. Both projen and jsii are structured around semver and the idea of immutable releases. It seems Maven Snapshots are quite different from that, a bit more like what a tag does on npmjs.com: always pointing to the latest version ("snapshot") of something.

**Reply by jrouly** [reply]
Yeah they definitely can be used that way, but I've also seen them used as immutable releases (which is how I'm using them).

My full package version is more like `{tag}-{sha}-{build number}-SNAPSHOT`, where snapshots are simply "non-production" releases. i.e., generated for every non-tagged commit.

**Comment by mrgrain** [comment]
Thinking a bit more about this, the easiest way to change the bumped version after the fact would be to `.spawn()` a task right after `bump`:

```ts
const fixVersions = project.addTask('fix-versions', {
  exec: "...",
});
project.tasks.tryFind('bump').spawn(fixVersions);
```

`fixVersions` would have to make sure to change all the required files (`package.json`, all files in `./dist`)

**Reply by jrouly** [reply]
Interesting suggestion! Yeah something like that could do the trick.... I'll see if I can work it in somehow.

**Comment by jrouly** [comment]
Actually I'm a dummy. I spent so long trying to get `npx projen bump` to work, I forgot I could just use `npm version prerelease` because the only thing that needs to change before `npx projen compile` runs is `package.json`, which `npm version prerelease` is perfectly capable of updating. 🤦
